### ISSUE TITLE
Naively: the default current working directory should be the project dir

### ISSUE THREAD

**Issue by kaste** [SublimeLinter/SublimeLinter#818]
This is coming from #581 

Current code for the `cwd` is

https://github.com/SublimeLinter/SublimeLinter/blob/ce7045121f6d32b7189ad9340921e5b285633867/lint/linter.py#L951-L962

which means we use the `dirname`, t.i. the folder, of the current file if we have a saved file. For unsaved files, we 'expand' the dot, which can be anything, e.g. it can be the sublime folder. 

I really don't know, what the rationale here is. For me, it seems unpredictable for unsaved, new files. It is not what I usually do and expect if I'm working on the command line. Everyones, almost all the time works from the root of the current project, to lint, run test commands, `npm build` whatever.

So, please enlighten me. Why is that? Should we change it?

**Comment by FichteFoll** [comment]
As an aside, I so something similar but more automatic for the mypy linter here:

https://github.com/fredcallaway/SublimeLinter-contrib-mypy/blob/cd6fbaf0cd9c8f52b23df621537fa9276a92bae1/linter.py#L134-L142

**Comment by pykong** [comment]
This!

**Comment by braver** [comment]
> So, please enlighten me. Why is that? Should we change it?

I tried a few things, and the result of `os.path.realpath('.')` never made any sense to me. I also don't see an API that gives us a more meaningful directory. So, I can totally get behind your proposal, at least that's predictable.

**Comment by FichteFoll** [comment]
Suggestion: Allow a linter-specific override for this as `cwd` that defaults to `${folder}`, which is either the project folder containing the file or the file's containing folder.

Highly related to #825.

**Comment by kaste** [comment]
`${folder:$file_path}` is exactly what I wanted to implement here. But this is blocked by #825 where I factored that `guess_project_path` fn. 

I don't think we need a fallback here, you proposed in #832, bc if we return `None` and pass it down to `popen`, it will not break anything, it will just not change the cwd, which is good.

We have the linter specific override, but the setting is called `chdir`. Plugin authors can provide a default
here, and end users can override this again in their settings. 

`chdir` would get the new var-expansions stuff from #825

**Comment by FichteFoll** [comment]
I think the complaint in this issue was also about `.` being used as the working dir, which is kind of unpredictable. I thought using the tempdir for that would be deterministic and more desirable. The fallback would only get used for untitled views anyway. I'm pretty much indifferent on this though.

**Comment by FichteFoll** [comment]
Also, since `chdir` is kind of weird when we don't really "change" the directory anymore, maybe we should rename the setting to `cwd`? I suspect that `chdir` was inferred from the underlying os module function.

**Comment by braver** [comment]
`chdir` is a command, `cwd` is a value. If we're setting a value we should call it `cwd`.

**Comment by kaste** [comment]
Hm 🤔 , a breaking change bc of name bike-shedding? It is now called `chdir` and end user would prefer maybe `working_dir` over `cwd` anyway. 

Or do you mean our method `get_chdir`? The setting is called `chdir`. The setting does not have to be called *current* working dir. It should have been *working_dir*. But that's bike-shedding.

If we break like in #825 than for a good reason. The old variable expansions were kind of unpredicatable and non-standard, we **add** a lot here. And we're still more readable, if you look at the source code. 

Anyway to make the proposal clear:

- The default *cwd* will be `${folder:$file_path}`. That is e.g. we always try - best effort here - to get a project path and use it. 

- For *unsaved* files, the project path will be the 'main' folder opened in a window. Or in code speak: `window.folders()[0]`.

- For saved files, we traverse all open folders to find a *best* match. The algorithm here can still be improved. See: https://github.com/kaste/SublimeLinter3/blob/9a42149a0c2c44fc4c24dfe6d844f4eb9947ac91/lint/linter.py#L510-L513

- If **no** folder is open, we take the folder (aka `os.path.dirname`) of the file in the view. 

- If **no** folder is open and we have an unsaved file, we do nothing. T.i. we do *not* chdir at all.

**Comment by braver** [comment]
> a breaking change bc of name bike-shedding

Although I think we *could* break settings, perhaps if we do so we should try to be backwards compatible. I don’t think this would carry enough benefits to go through the trouble though, so 🤷🏼‍♂️

Proposal looks good!

**Comment by FichteFoll** [comment]
Well, since we are breaking the entire configuration anyway, this seems like a good opportunity to make such a change. I'm all in on `working_dir` over `cwd` as that's more explicit and conveys the meaning better. It's also the name used in ST's build systems.

The method name would then be adjusted as well, but that's most likely not breaking as its primarily used internally.

**Comment by braver** [comment]
I can get behind that. It's a whole different SublimeLinter from a settings point of view anyway. If you guys wanna go there, refactor and make it more sane I'm game.

**Comment by kaste** [comment]
Only @FichteFoll's *mypy* is using `get_chdir` 😜

**Comment by FichteFoll** [comment]
```
~/code/sublimelinter_channel Σ rg "chdir" linter_repos linter_repos-contrib
linter_repos/SublimeLinter-eslint/README.md
140:If you use a SublimeText project and `.sublime-project` file is in project folder, set `chdir` to `${project}` in your SublimeLinter settings. **Warning:** it can cause bugs if your project has more than one root folder.
linter_repos-contrib/SublimeLinter-contrib-rustc/linter.py
76:                os.chdir(os.path.dirname(self.cargo_config))
85:                    os.chdir(old_cwd)
linter_repos-contrib/SublimeLinter-contrib-mypy/linter.py
118:    def get_chdir(self, settings):
119:        """Find the chdir to use with the linter."""
125:        # Users could set the `chdir` linter setting,
128:        chdir = settings.get('chdir', None)
130:        if chdir and os.path.isdir(chdir):
131:            persist.debug('chdir has been set to: {0}'.format(chdir))
132:            return chdir
135:                # Only difference from Linter.get_chdir
136:                chdir = _find_first_nonpackage_parent(self.filename)
137:                if chdir != os.path.dirname(self.filename):
138:                    persist.debug('chdir has been set to: {0}'.format(chdir))
139:                return chdir
linter_repos-contrib/SublimeLinter-contrib-elixirc/linter.py
257:            filename = os.path.join(self.chdir, dummy_match.group('filename'))
linter_repos-contrib/SublimeLinter-contrib-elixirc/README.md
38:| chdir       | _(string)_ run linter from the specified dir     |
43:* set `chdir` to your mix project's root directory.
53:            "chdir": "PROJECT_ROOT"
102:          "chdir": "/Users/foo/projects/myapp",
linter_repos-contrib/SublimeLinter-contrib-codeclimate/linter.py
42:        'chdir': "${project}"
linter_repos-contrib/SublimeLinter-contrib-elm-make/linter.py
48:            os.chdir(root_dir)
linter_repos-contrib/SublimeLinter-contrib-markdownlint/README.md
69:[`chdir` setting] to ensure `markdownlint` is executed relative to your
107:[`chdir` setting]: http://sublimelinter.readthedocs.io/en/latest/linter_settings.html#chdir
linter_repos-contrib/SublimeLinter-contrib-elixirc/messages/1.3.0.txt
18:          "chdir": "/Users/foo/projects/myapp",
```

**Comment by kaste** [comment]
What does that mean 🤷‍♂️ You forgot to comment?

**Comment by FichteFoll** [comment]
True, I should've left a comment.

The above post shows all instances of `chdir` being used as either the function or the setting in current linters.

The `chdir` default settigs we want to deprecate by providing that as a default in the first place,
so we can ignore that. What's left are:

- **mypy**: Will need to consider whether project dir or "first dir without an `__init__.py` makes more sense - can probably just rip this out as using the project dir should cover most cases.
- **rustc**: Uses `os.chdir` directly after determining the project path from the presence of a `Cargo.toml` file, probably. This will need some special attention.
- **elixirc**: Accesses `self.chdir` (is that even set anywhere?)

**Comment by kaste** [comment]
**elixir** `self.chdir` is not a thing.

**rustc** this code is not thread safe, and already in trouble. Our popen patch from last week would make it easy to make their code thread safe. Relatively unrelated to this issue here though.

**mypy** is under your control

> linter_repos-contrib/SublimeLinter-contrib-codeclimate/linter.py
42:        'chdir': "${project}"

Probably already broken bc ${project} now means something different. Also not related to this here.

**Comment by FichteFoll** [comment]
Agreed. I'll probably just remove the special handling from mypy. Rustc uses os.chdir directly, so it's unrelated to this issue, and all the other occurrences are already wrong or meant to do something we want to turn into the default anyway.

I say we can rename this to `working_dir` safely.
